nimble/host/audio/src/ble_audio.c

```diff
@@ -0,0 +1,524 @@
+        SLIST_INSERT_HEAD(&ble_audio_event_listener_list, listener, next);
+        rc = 0;
+    } else {
+        rc = BLE_HS_EALREADY;
+    }
+
+    return rc;
+}
+
+int
+ble_audio_event_listener_unregister(struct ble_audio_event_listener *listener)
+{
+    struct ble_audio_event_listener *evl = NULL;
+    int rc;
+
+    if (listener == NULL) {
+        BLE_HS_LOG_ERROR("NULL listener!\n");
+        return BLE_HS_EINVAL;
+    }
+
+    /*
+     * We check if element exists on the list only for sanity to let caller
+     * know whether it registered its listener before.
+     */
+
+    SLIST_FOREACH(evl, &ble_audio_event_listener_list, next) {
+        if (evl == listener) {
+            break;
+        }
+    }
+
+    if (!evl) {
+        rc = BLE_HS_ENOENT;
+    } else {
+        SLIST_REMOVE(&ble_audio_event_listener_list, listener,
+                     ble_audio_event_listener, next);
+
+        if (SLIST_EMPTY(&ble_audio_event_listener_list)) {
+            rc = ble_gap_event_listener_unregister(
+                    &ble_audio_gap_event_listener);
+        } else {
+            rc = 0;
+        }
+    }
+
+    return rc;
+}
+int
+ble_audio_event_listener_call(struct ble_audio_event *event)
+{
+    struct ble_audio_event_listener *evl = NULL;
+
+    SLIST_FOREACH(evl, &ble_audio_event_listener_list, next) {
+        evl->fn(event, evl->arg);
+    }
+
+    return 0;
+}
+
+/* Get the next subgroup data pointer */
+static const uint8_t *
+ble_audio_base_subgroup_next(uint8_t num_bis, const uint8_t *data,
+                             uint8_t data_len)
+{
+    uint8_t offset = 0;
+
+    for (uint8_t i = 0; i < num_bis; i++) {
+        uint8_t codec_specific_config_len;
+
+        /* BIS_index[i[k]] + Codec_Specific_Configuration_Length[i[k]] */
+        if ((data_len - offset) < 2) {
+            return NULL;
+        }
+
+        /* Skip BIS_index[i[k]] */
+        offset++;
+
+        codec_specific_config_len = data[offset];
+        offset++;
+
+        if ((data_len - offset) < codec_specific_config_len) {
+            return NULL;
+        }
+
+        offset += codec_specific_config_len;
+    }
+
+    return &data[offset];
+}
+
+int
+ble_audio_base_parse(const uint8_t *data, uint8_t data_len,
+                     struct ble_audio_base_group *group,
+                     struct ble_audio_base_iter *subgroup_iter)
+{
+    uint8_t offset = 0;
+
+    if (data == NULL) {
+        BLE_HS_LOG_ERROR("NULL data!\n");
+        return BLE_HS_EINVAL;
+    }
+
+    if (group == NULL) {
+        BLE_HS_LOG_ERROR("NULL group!\n");
+        return BLE_HS_EINVAL;
+    }
+
+    /* Presentation_Delay + Num_Subgroups */
+    if (data_len < 4) {
+        return BLE_HS_EMSGSIZE;
+    }
+
+    group->presentation_delay = get_le24(data);
+    offset += 3;
+
+    group->num_subgroups = data[offset];
+    offset++;
+
+    if (group->num_subgroups < 1) {
+        BLE_HS_LOG_DEBUG("Rule 1 violation: There shall be at least one subgroup.\n");
+    }
```

Comment:
We know the rules, but it imho looks a bit weird :) And don't we want to return early here, like in other errors? I think BASE with no subgroups as just as bad as empty one - both don't follow spec. Maybe something like this?
```suggestion
    if (group->num_subgroups < 1) {
        BLE_HS_LOG_ERROR("Invalid BASE: no subgroups!\n");
        return BLE_HS_EINVAL;
    }
```

Reply:
Yeah, I was considering that, probably you're right.
I was thinking about this as an Advertising data that display to the user - when you parse e.g. Complete Name, Service Data and Appearance. When you fail to parse Service data I think you still want to report the Complete Name and Appearance.
Let's wait for 3rd eye pair to look at this.

Reply:
I suggest BLE_HS_LOG_ERROR here and return as @KKopyscinski  suggested.

Reply:
I like @MariuszSkamra  error message better, so just change DEBUG to ERROR here.